### PR TITLE
span: set error tag based on span kind and status code

### PR DESCRIPTION
**Pull request by lonnblad**

### CODE REVIEW

**Comment by lonnblad** [comment]
fails on
```
--- FAIL: TestTraceExporter (0.00s)
    --- FAIL: TestTraceExporter/threshold (0.00s)
        span_test.go:452: mismatch: got 4, wanted 5
```

should this be fixed or should the number be changed to 4?
from the code: `// 5 spans should take us overboard`

**Comment by cgilmour** [comment]
The span definition got changed to have a tiny bit more data.
https://github.com/DataDog/opencensus-go-exporter-datadog/blob/8cb3a4189560015e48cf39e9c3dbd85d87d2cbae/span_test.go#L276-L279
So the test now only needs 4 spans to reach the "flush threshold" instead of 5. The `count` value should be changed.

The `equalFunc` in span_test.go should also mark the function as a helper, so a useful line number is reported, eg:
```
func equalFunc(t *testing.T) func(got, want interface{}) {
        return func(a, b interface{}) {
                t.Helper()
                if !reflect.DeepEqual(a, b) {
```
Then it should report `trace_test.go:75` instead of `span_test.go:452`

File: span.go [hunk 1]
```diff
@@ -8,39 +8,52 @@ package datadog
...
+	trace.StatusCodeInternal:           {message: "internal", httpCode: http.StatusInternalServerError},
+	trace.StatusCodeUnavailable:        {message: "unavailable", httpCode: http.StatusServiceUnavailable},
+	trace.StatusCodeDataLoss:           {message: "data_loss", httpCode: http.StatusNotImplemented},
+	trace.StatusCodeUnauthenticated:    {message: "unauthenticated", httpCode: http.StatusUnauthorized},
 }
 
-func canonicalCodeString(code int32) string {
-	if code < 0 || int(code) >= len(canonicalCodes) {
-		return "error code " + strconv.FormatInt(int64(code), 10)
+type statusCode struct {
```
Comment (gbbr) [inline span.go]: Let's call this `codeDetails` and please add some documentation. e.g.
```go
// codeDetails specifies information about a trace status code.
type codeDetails struct {
    message string // status message
    status  int    // corresponding HTTP status code
}
```
It seems like a more appropriate name.

File: span.go [hunk 2]
```diff
@@ -8,39 +8,52 @@ package datadog
...
-	"failed_precondition",
-	"aborted",
-	"out_of_range",
-	"unimplemented",
-	"internal",
-	"unavailable",
-	"data_loss",
-	"unauthenticated",
+var statusCodes = map[int32]statusCode{
+	trace.StatusCodeOK:                 {message: "ok", httpCode: http.StatusOK},
```
Comment (gbbr) [inline span.go]: Should we capitalize all the messages? I'm thinking to keep consistent with: https://github.com/census-instrumentation/opencensus-go/blob/b4a14686f0a98096416fe1b4cb848e384fb2b22b/plugin/ochttp/trace.go#L212-L228

File: span.go [hunk 3]
```diff
@@ -8,39 +8,52 @@ package datadog
...
+
+func statusMessage(code int32) string {
+	if sc, exists := statusCodes[code]; exists {
+		return sc.message
 	}
-	return canonicalCodes[code]
+	return "error code " + strconv.FormatInt(int64(code), 10)
+}
+
+func httpCode(code int32) int {
```
Comment (gbbr) [inline span.go]: Same here. This function is also not really needed.

File: span.go [hunk 4]
```diff
@@ -8,39 +8,52 @@ package datadog
...
 
-func canonicalCodeString(code int32) string {
-	if code < 0 || int(code) >= len(canonicalCodes) {
-		return "error code " + strconv.FormatInt(int64(code), 10)
+type statusCode struct {
+	message  string
+	httpCode int
+}
+
+func statusMessage(code int32) string {
```
Comment (gbbr) [inline span.go]: I think this function is not necessary at this point.

File: span.go [hunk 5]
```diff
@@ -60,21 +73,38 @@ func (e *traceExporter) convertSpan(s *trace.SpanData) *ddSpan {
 	if s.ParentSpanID != (trace.SpanID{}) {
 		span.ParentID = binary.BigEndian.Uint64(s.ParentSpanID[:])
 	}
+
+	httpCode := httpCode(s.Status.Code)
```
Comment (gbbr) [inline span.go]: Can we call this simply `code`? Then, we can just do:
```go
code, ok := statusCodes[s.Status.Code]
if !ok {
    code = codeDetails{
        message: "ERR_CODE_" + strconv.FormatInt(int64(s.Status.Code), 10),
        status: http.StatusInternalServerError,
    }
}
```
Then we can just use `code.message` and `code.status` everywhere.

Or, if you really prefer to have a function, we could just have it return `codeDetails` directly, no need to have two separate functions.

File: span.go [hunk 6]
```diff
@@ -60,21 +73,38 @@ func (e *traceExporter) convertSpan(s *trace.SpanData) *ddSpan {
...
 	switch s.SpanKind {
 	case trace.SpanKindClient:
 		span.Type = "client"
+		if httpCode >= 400 && httpCode < 500 {
+			span.Error = 1
+		}
 	case trace.SpanKindServer:
 		span.Type = "server"
+		if httpCode >= 500 && httpCode < 600 {
```
Comment (gbbr) [inline span.go]: I think you can remove this condition and just use `fallthrough`, right?

File: span.go [hunk 7]
```diff
@@ -60,21 +73,38 @@ func (e *traceExporter) convertSpan(s *trace.SpanData) *ddSpan {
 	if s.ParentSpanID != (trace.SpanID{}) {
 		span.ParentID = binary.BigEndian.Uint64(s.ParentSpanID[:])
 	}
+
+	httpCode := httpCode(s.Status.Code)
 	switch s.SpanKind {
 	case trace.SpanKindClient:
 		span.Type = "client"
+		if httpCode >= 400 && httpCode < 500 {
```
Comment (gbbr) [inline span.go]: I'm wondering if it's nicer and easier to reason about as:
```go
if code/100 == 4 {
```
🤔

File: span.go [hunk 8]
```diff
@@ -8,39 +8,52 @@ package datadog
...
+	trace.StatusCodeDataLoss:           {message: "data_loss", httpCode: http.StatusNotImplemented},
+	trace.StatusCodeUnauthenticated:    {message: "unauthenticated", httpCode: http.StatusUnauthorized},
 }
 
-func canonicalCodeString(code int32) string {
-	if code < 0 || int(code) >= len(canonicalCodes) {
-		return "error code " + strconv.FormatInt(int64(code), 10)
+type statusCode struct {
+	message  string
+	httpCode int
```
Comment (gbbr) [inline span.go]: Can we simply call this `status`?

File: span.go (hunk 2 above)
Comment (lonnblad) [reply span.go]: sounds good!

do you think that statusCodes is a good name?
the proto file says: `The canonical error codes for Google APIs.`

File: span.go (hunk 6 above)
Comment (lonnblad) [reply span.go]: yep

File: span.go (hunk 7 above)
Comment (lonnblad) [reply span.go]: 👍

**Comment by lonnblad** [comment]
@gbbr thanks for the comments, I made some updates! :)

**Comment by gbbr** [comment]
Thanks @lonnblad. This LGTM. Would love a pair of eyes from @cgilmour too.

**Pull request merged**